e2e_tests/tests/experiment/test_custom_searcher.py

```diff
@@ -70,3 +71,114 @@ def initial_operations(self) -> List[Operation]:
         close = Close(request_id=create.request_id)
         logging.debug(f"Create({create.request_id}, {create.hparams})")
         return [create, validate_after, close]
+
+
+@pytest.mark.e2e_cpu
+def test_run_random_searcher_exp() -> None:
+    config = conf.load_config(conf.fixtures_path("no_op/single.yaml"))
+    config["searcher"] = {
+        "name": "custom",
+        "metric": "validation_error",
+        "smaller_is_better": True,
+        "max_length": {"batches": 3000},
+    }
+    config["description"] = "custom searcher"
+
+    max_trials = 10
+    max_concurrent_trials = 1
+
+    search_method = RandomSearcherMethod(max_trials, max_concurrent_trials)
+    search_runner = SearchRunner(search_method)
+    search_runner.run(config, context_dir=conf.fixtures_path("no_op"))
+
+
+class RandomSearcherMethod(SearchMethod):
+    def __init__(self, max_trials: int, max_concurrent_trials: int) -> None:
+        super().__init__(SearchState(None))
+        self.max_trials = max_trials
+        self.max_concurrent_trials = max_concurrent_trials
+
+        self.created_trials = 0
+        self.pending_trials = 0
+        self.closed_trials = 0
+
+    def on_trial_created(self, request_id: uuid.UUID) -> List[Operation]:
+        self._log_stats()
+        return []
+
+    def on_validation_completed(self, request_id: uuid.UUID, metric: float) -> List[Operation]:
+        return []
+
+    def on_trial_closed(self, request_id: uuid.UUID) -> List[Operation]:
+        self.pending_trials -= 1
+        self.closed_trials += 1
+        ops = []
+        if self.created_trials < self.max_trials:
+            request_id = uuid.uuid4()
+            ops.append(Create(request_id=request_id, hparams=self.sample_params(), checkpoint=None))
+            ops.append(ValidateAfter(request_id=request_id, length=3000))
+            ops.append(Close(request_id=request_id))
+            self.created_trials += 1
+            self.pending_trials += 1
```

Comment:
```suggestion
        if self.created_trials < self.max_trials:
            request_id = uuid.uuid4()
            ops.append(Create(request_id=request_id, hparams=self.sample_params(), checkpoint=None))
            ops.append(ValidateAfter(request_id=request_id, length=3000))
            ops.append(Close(request_id=request_id))
            self.created_trials += 1
            self.pending_trials += 1
        else:
            ops.append(Shutdown())
```

Reply:
Updated the code to invoke Shutdown the same way as in [Nikita's PR ](https://github.com/determined-ai/determined/pull/4668/files#:~:text=%7D%20else%20if%20s.PendingTrials%20%3D%3D%200%20%7B%20//%20even,ops%20%3D%20append(ops%2C%20shutdown)).
We may still need to add it in `trail_exited_early` though. Checking now.